make ForgedAlliance.exe executable on Linux

when starting a game on Linux the following error occures: 
```
2019-01-02 13:00:26.006  WARN 7291 --- [     Thread-627] com.faforever.client.game.GameService    : Game could not be started

java.util.concurrent.CompletionException: com.github.nocatch.NoCatchException: java.io.IOException: Cannot run program "~/.faforever/bin/ForgedAlliance.exe" (in directory "~/.faforever/bin"): error=13, Permission denied
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[na:na]
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[na:na]
	at com.faforever.client.replay.ReplayServerImpl.lambda$start$2(ReplayServerImpl.java:109) ~[downlords-faf-client-0.9.3-beta.jar:0.9.3-beta]
	at java.base/java.lang.Thread.run(Thread.java:844) ~[na:na]
Caused by: com.github.nocatch.NoCatchException: java.io.IOException: Cannot run program "/home/sws/.faforever/bin/ForgedAlliance.exe" (in directory "/home/sws/.faforever/bin"): error=13, Permission denied
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:na]
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:488) ~[na:na]
	at com.github.nocatch.NoCatch.wrapException(NoCatch.java:61) ~[nocatch-1.1.jar:na]
	at com.github.nocatch.NoCatch.noCatch(NoCatch.java:104) ~[nocatch-1.1.jar:na]
	at com.github.nocatch.NoCatch.noCatch(NoCatch.java:82) ~[nocatch-1.1.jar:na]
	at com.faforever.client.game.GameService.lambda$startGame$33(GameService.java:467) ~[downlords-faf-client-0.9.3-beta.jar:0.9.3-beta]
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:714) ~[na:na]
	... 4 common frames omitted
Caused by: java.io.IOException: Cannot run program "/home/sws/.faforever/bin/ForgedAlliance.exe" (in directory "/home/sws/.faforever/bin"): error=13, Permission denied
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128) ~[na:na]
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071) ~[na:na]
	at com.faforever.client.fa.ForgedAllianceService.launch(ForgedAllianceService.java:128) ~[downlords-faf-client-0.9.3-beta.jar:0.9.3-beta]
	at com.faforever.client.fa.ForgedAllianceService.startGame(ForgedAllianceService.java:73) ~[downlords-faf-client-0.9.3-beta.jar:0.9.3-beta]
	at com.faforever.client.game.GameService.lambda$startGame$32(GameService.java:467) ~[downlords-faf-client-0.9.3-beta.jar:0.9.3-beta]
	at com.github.nocatch.NoCatch.noCatch(NoCatch.java:100) ~[nocatch-1.1.jar:na]
	... 7 common frames omitted
Caused by: java.io.IOException: error=13, Permission denied
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method) ~[na:na]
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:339) ~[na:na]
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:270) ~[na:na]
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107) ~[na:na]
	... 12 common frames omitted
```

Setting the executable bit on `~/.faforever/bin/ForgedAlliance.exe` fixes this error

--- a/src/main/java/com/faforever/client/fx/PlatformService.java
+++ b/src/main/java/com/faforever/client/fx/PlatformService.java
@@ -137,8 +137,8 @@ public class PlatformService {
     return windowTitle.equals(getForegroundWindowTitle());
   }
 
-  public void setUnixExecutableBit(Path exePath) throws IOException {
-	// client needs executable bit for running and the writeable bit set to alter the version
+  public void setUnixExecutableAndWritableBits(Path exePath) throws IOException {
+    // client needs executable bit for running and the writeable bit set to alter the version
     if (SystemUtils.IS_OS_UNIX) {
       Files.setPosixFilePermissions(exePath, Sets.immutableEnumSet(PosixFilePermission.OWNER_READ,
     				  PosixFilePermission.OWNER_WRITE, PosixFilePermission.OWNER_EXECUTE));

--- a/src/main/java/com/faforever/client/fx/PlatformService.java
+++ b/src/main/java/com/faforever/client/fx/PlatformService.java
@@ -1,14 +1,20 @@
 package com.faforever.client.fx;
 
+import com.google.common.collect.Sets;
 import com.sun.jna.Platform;
 import com.sun.jna.platform.win32.User32;
 import com.sun.jna.platform.win32.WinDef.HWND;
 import com.sun.jna.platform.win32.WinUser;
 import com.sun.jna.platform.win32.WinUser.WINDOWPLACEMENT;
 import javafx.application.HostServices;
+
+import org.apache.commons.lang3.SystemUtils;
 import org.jetbrains.annotations.Nullable;
 
+import java.io.IOException;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermission;
 
 import static com.github.nocatch.NoCatch.noCatch;
 import static org.bridj.Platform.show;
@@ -130,4 +136,13 @@ public class PlatformService {
   public boolean isWindowFocused(String windowTitle) {
     return windowTitle.equals(getForegroundWindowTitle());
   }
+
+  public void setUnixExecutableBit(Path exePath) throws IOException {
+	// client needs executable bit for running and the writeable bit set to alter the version
+    if (SystemUtils.IS_OS_UNIX) {
+      Files.setPosixFilePermissions(exePath, Sets.immutableEnumSet(PosixFilePermission.OWNER_READ,
+    				  PosixFilePermission.OWNER_WRITE, PosixFilePermission.OWNER_EXECUTE));
+    }
+  }
 }
+

--- a/src/main/java/com/faforever/client/patch/GameBinariesUpdateTaskImpl.java
+++ b/src/main/java/com/faforever/client/patch/GameBinariesUpdateTaskImpl.java
@@ -1,6 +1,7 @@
 package com.faforever.client.patch;
 
 import com.faforever.client.config.ClientProperties;
+import com.faforever.client.fx.PlatformService;
 import com.faforever.client.i18n.I18n;
 import com.faforever.client.preferences.PreferencesService;
 import com.faforever.client.task.CompletableTask;
@@ -10,6 +11,7 @@ import com.faforever.client.util.Validator;
 import com.faforever.commons.fa.ForgedAllianceExePatcher;
 import com.faforever.commons.io.ByteCopier;
 import com.google.common.annotations.VisibleForTesting;
+
 import org.apache.maven.artifact.versioning.ComparableVersion;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -19,7 +21,6 @@ import org.springframework.stereotype.Component;
 
 import javax.inject.Inject;
 
-import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
@@ -63,17 +64,19 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
 
   private final I18n i18n;
   private final PreferencesService preferencesService;
+  private final PlatformService platformService;
 
   private final String fafExeUrl;
 
   private Integer version;
 
   @Inject
-  public GameBinariesUpdateTaskImpl(I18n i18n, PreferencesService preferencesService, ClientProperties clientProperties) {
+  public GameBinariesUpdateTaskImpl(I18n i18n, PreferencesService preferencesService, PlatformService platformService, ClientProperties clientProperties) {
     super(Priority.HIGH);
 
     this.i18n = i18n;
     this.preferencesService = preferencesService;
+    this.platformService = platformService;
 
     this.fafExeUrl = clientProperties.getForgedAlliance().getExeUrl();
   }
@@ -95,7 +98,7 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
 
   private void downloadFafExeIfNecessary(Path exePath) throws IOException {
     if (Files.exists(exePath)) {
-      makeSureExeIsExecutable(exePath);
+      platformService.setUnixExecutableBit(exePath);
       return;
     }
     ResourceLocks.acquireDownloadLock();
@@ -110,14 +113,10 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
             .listener(this::updateProgress)
             .copy();
       }
-      makeSureExeIsExecutable(exePath);
+      platformService.setUnixExecutableBit(exePath);
     } finally {
       ResourceLocks.freeDownloadLock();
     }
-  }
-
-  private void makeSureExeIsExecutable(Path exePath) {
-    exePath.toFile().setExecutable(true);
   }
 
   @VisibleForTesting

--- a/src/main/java/com/faforever/client/patch/GameBinariesUpdateTaskImpl.java
+++ b/src/main/java/com/faforever/client/patch/GameBinariesUpdateTaskImpl.java
@@ -18,6 +18,8 @@ import org.springframework.context.annotation.Scope;
 import org.springframework.stereotype.Component;
 
 import javax.inject.Inject;
+
+import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
@@ -93,6 +95,7 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
 
   private void downloadFafExeIfNecessary(Path exePath) throws IOException {
     if (Files.exists(exePath)) {
+      makeSureExeIsExecutable(exePath);
       return;
     }
     ResourceLocks.acquireDownloadLock();
@@ -107,9 +110,14 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
             .listener(this::updateProgress)
             .copy();
       }
+      makeSureExeIsExecutable(exePath);
     } finally {
       ResourceLocks.freeDownloadLock();
     }
+  }
+
+  private void makeSureExeIsExecutable(Path exePath) {
+    exePath.toFile().setExecutable(true);
   }
 
   @VisibleForTesting

--- a/src/main/java/com/faforever/client/patch/GameBinariesUpdateTaskImpl.java
+++ b/src/main/java/com/faforever/client/patch/GameBinariesUpdateTaskImpl.java
@@ -96,9 +96,10 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
     return null;
   }
 
-  private void downloadFafExeIfNecessary(Path exePath) throws IOException {
+  @VisibleForTesting
+  void downloadFafExeIfNecessary(Path exePath) throws IOException {
     if (Files.exists(exePath)) {
-      platformService.setUnixExecutableBit(exePath);
+      platformService.setUnixExecutableAndWritableBits(exePath);
       return;
     }
     ResourceLocks.acquireDownloadLock();
@@ -113,7 +114,7 @@ public class GameBinariesUpdateTaskImpl extends CompletableTask<Void> implements
             .listener(this::updateProgress)
             .copy();
       }
-      platformService.setUnixExecutableBit(exePath);
+      platformService.setUnixExecutableAndWritableBits(exePath);
     } finally {
       ResourceLocks.freeDownloadLock();
     }

--- a/src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java
+++ b/src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java
@@ -95,7 +95,12 @@ public class ClientUpdateServiceImpl implements ClientUpdateService {
   }
 
   private void install(Path binaryPath) {
-    // TODO probably need to make this executable on unix See #1026
+    // need to make this executable on unix See #1026
+	try {
+      platformService.setUnixExecutableBit(binaryPath);
+	} catch (IOException e) {
+      logger.warn("Unix execute bit could not be set", e);
+	}
     String command = binaryPath.toAbsolutePath().toString();
     try {
       logger.info("Starting installer at {}", command);

--- a/src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java
+++ b/src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java
@@ -94,13 +94,13 @@ public class ClientUpdateServiceImpl implements ClientUpdateService {
     return currentVersion;
   }
 
-  private void install(Path binaryPath) {
-    // need to make this executable on unix See #1026
-	try {
-      platformService.setUnixExecutableBit(binaryPath);
-	} catch (IOException e) {
+  @VisibleForTesting
+  void install(Path binaryPath) {
+  	try {
+      platformService.setUnixExecutableAndWritableBits(binaryPath);
+  	} catch (IOException e) {
       logger.warn("Unix execute bit could not be set", e);
-	}
+  	}
     String command = binaryPath.toAbsolutePath().toString();
     try {
       logger.info("Starting installer at {}", command);

--- a/src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java
+++ b/src/main/java/com/faforever/client/update/ClientUpdateServiceImpl.java
@@ -96,11 +96,11 @@ public class ClientUpdateServiceImpl implements ClientUpdateService {
 
   @VisibleForTesting
   void install(Path binaryPath) {
-  	try {
+    try {
       platformService.setUnixExecutableAndWritableBits(binaryPath);
-  	} catch (IOException e) {
+    } catch (IOException e) {
       logger.warn("Unix execute bit could not be set", e);
-  	}
+    }
     String command = binaryPath.toAbsolutePath().toString();
     try {
       logger.info("Starting installer at {}", command);

--- a/src/test/java/com/faforever/client/patch/GameBinariesUpdateTaskTest.java
+++ b/src/test/java/com/faforever/client/patch/GameBinariesUpdateTaskTest.java
@@ -10,6 +10,7 @@ import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TemporaryFolder;
 import org.mockito.Mock;
+import org.mockito.Mockito;
 import org.mockito.MockitoAnnotations;
 
 import java.io.IOException;
@@ -18,6 +19,7 @@ import java.nio.file.Path;
 import java.util.List;
 import java.util.stream.Collectors;
 
+import static com.faforever.client.preferences.PreferencesService.FORGED_ALLIANCE_EXE;
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
@@ -77,6 +79,13 @@ public class GameBinariesUpdateTaskTest {
     for (String fileName : GameBinariesUpdateTaskImpl.BINARIES_TO_COPY) {
       assertTrue(java.nio.file.Files.exists(fafBinPath.resolve(fileName)));
     }
+  }
+
+  @Test
+  public void testUnixExecutableBitIsSet() throws Exception {
+    Path faExePath = fafBinDirectory.newFile("ForgedAlliance.exe").toPath();
+    instance.downloadFafExeIfNecessary(faExePath);
+    Mockito.verify(platformService).setUnixExecutableAndWritableBits(faExePath);
   }
 
   private void createFileWithSize(Path file, int size) throws IOException {

--- a/src/test/java/com/faforever/client/patch/GameBinariesUpdateTaskTest.java
+++ b/src/test/java/com/faforever/client/patch/GameBinariesUpdateTaskTest.java
@@ -1,6 +1,7 @@
 package com.faforever.client.patch;
 
 import com.faforever.client.config.ClientProperties;
+import com.faforever.client.fx.PlatformService;
 import com.faforever.client.i18n.I18n;
 import com.faforever.client.preferences.Preferences;
 import com.faforever.client.preferences.PreferencesService;
@@ -30,6 +31,8 @@ public class GameBinariesUpdateTaskTest {
   @Mock
   private PreferencesService preferencesService;
   @Mock
+  private PlatformService platformService;
+  @Mock
   private I18n i18n;
 
   private GameBinariesUpdateTaskImpl instance;
@@ -38,7 +41,7 @@ public class GameBinariesUpdateTaskTest {
   public void setUp() throws Exception {
     MockitoAnnotations.initMocks(this);
 
-    instance = new GameBinariesUpdateTaskImpl(i18n, preferencesService, new ClientProperties());
+    instance = new GameBinariesUpdateTaskImpl(i18n, preferencesService, platformService, new ClientProperties());
 
     Path faPath = faDirectory.getRoot().toPath();
     java.nio.file.Files.createDirectories(faPath.resolve("bin"));

--- a/src/test/java/com/faforever/client/update/ClientUpdateServiceImplTest.java
+++ b/src/test/java/com/faforever/client/update/ClientUpdateServiceImplTest.java
@@ -8,14 +8,18 @@ import com.faforever.client.task.TaskService;
 import com.faforever.commons.io.Bytes;
 import org.apache.maven.artifact.versioning.ComparableVersion;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
 import org.junit.runner.RunWith;
 import org.mockito.ArgumentCaptor;
 import org.mockito.Mock;
+import org.mockito.Mockito;
 import org.mockito.junit.MockitoJUnitRunner;
 import org.springframework.context.ApplicationContext;
 
 import java.net.URL;
+import java.nio.file.Path;
 import java.util.concurrent.CompletableFuture;
 
 import static com.faforever.client.notification.Severity.INFO;
@@ -32,6 +36,8 @@ public class ClientUpdateServiceImplTest {
 
   private ClientUpdateServiceImpl instance;
 
+  @Rule
+  public TemporaryFolder fafBinDirectory = new TemporaryFolder();
   @Mock
   private NotificationService notificationService;
   @Mock
@@ -76,5 +82,12 @@ public class ClientUpdateServiceImplTest {
 
     verify(i18n).get("clientUpdateAvailable.notification", "v0.4.8.1-alpha", Bytes.formatSize(56079360L, i18n.getUserSpecificLocale()));
     assertThat(persistentNotification.getSeverity(), is(INFO));
+  }
+
+  @Test
+  public void testUnixExecutableBitIsSet() throws Exception {
+    Path faExePath = fafBinDirectory.newFile("ForgedAlliance.exe").toPath();
+    instance.install(faExePath);
+    Mockito.verify(platformService).setUnixExecutableAndWritableBits(faExePath);
   }
 }

